prover/src/gpu.rs

```diff
@@ -0,0 +1,284 @@
+
+        // generate hashes on the gpu
+        let lde_domain_size = domain.lde_domain_size();
+        let num_base_columns = trace.num_base_cols();
+        let rpo_requires_padding = num_base_columns % RPO_RATE != 0;
+        let rpo_padded_segment_idx = rpo_requires_padding.then_some(num_base_columns / RPO_RATE);
+        let mut row_hasher = GpuRpo256RowMajor::<Felt>::new(lde_domain_size, rpo_requires_padding);
+        let mut rpo_padded_segment: Vec<[Felt; RPO_RATE]>;
+        let mut lde_segment_iter = lde_segment_generator.gen_segment_iter().enumerate();
+        for (segment_idx, segment) in &mut lde_segment_iter {
+            let segment = lde_segments.push_get(Box::new(segment));
+            // check if the segment requires padding
+            if rpo_padded_segment_idx.map_or(false, |pad_idx| pad_idx == segment_idx) {
+                // duplicate and modify the last segment with Rpo256's padding
+                // rule ("1" followed by "0"s). Our segments are already
+                // padded with "0"s we only need to add the "1"s.
+                let rpo_pad_column = num_base_columns % RPO_RATE;
+                rpo_padded_segment = unsafe { page_aligned_uninit_vector(domain_size) };
+                rpo_padded_segment.copy_from_slice(segment);
+                rpo_padded_segment.iter_mut().for_each(|row| row[rpo_pad_column] = Felt::ONE);
+                row_hasher.update(&rpo_padded_segment);
+                assert!(lde_segment_iter.next().is_none(), "padded segment should be the last");
+                break;
+            }
+            row_hasher.update(segment);
+        }
+        let row_hashes = block_on(row_hasher.finish());
+        let tree_nodes = gen_rpo_merkle_tree(&row_hashes);
+        // aggregate segments at the same time as the GPU generates the merkle tree nodes
+        let lde_segments = lde_segments.into_vec().into_iter().map(|p| *p).collect();
+        let trace_lde = RowMatrix::from_segments(lde_segments, num_base_columns);
+        let trace_polys = lde_segment_generator.into_polys().unwrap();
+        let nodes = block_on(tree_nodes).into_iter().map(RpoDigest::new).collect();
+        let leaves = row_hashes.into_iter().map(RpoDigest::new).collect();
+        let trace_tree = MerkleTree::from_raw_parts(nodes, leaves).unwrap();
+        debug!(
+            "Extended (on CPU) and committed (on GPU) to an execution trace of {} columns from 2^{} to 2^{} steps in {} ms",
+            trace_polys.num_cols(),
+            trace_polys.num_rows().ilog2(),
+            trace_lde.num_rows().ilog2(),
+            now.elapsed().as_millis()
+        );
+
+        (trace_lde, trace_tree, trace_polys)
+    }
+}
+
+struct SegmentGenerator<'a, 'b, E, I, const N: usize>
+where
+    E: FieldElement<BaseField = Felt>,
+    I: IntoIterator<Item = Vec<E>>,
+{
+    poly_iter: I::IntoIter,
+    polys: Option<ColMatrix<E>>,
+    poly_offset: usize,
+    offsets: &'a [E::BaseField],
+    twiddles: &'b [E::BaseField],
+}
```

Comment:
One thing I didn't understand: why do we need both `poly_iter` and `polys` here? Would having a reference to `tracy_polys` built on line 82 above not be sufficient?

Also, if we pass `StarkDomain` via the constructor, we could keep a reference to it and build `offsets` internally. Maybe the struct could look something like this?
```Rust
struct SegmentGenerator<'a, E, const N: usize>
where
    E: FieldElement<BaseField = Felt>,
{
    trace_polys: &'a ColMatrix<E>,
    poly_offset: usize,
    offsets: Vec<E::BaseField>,
    domain: &'a StarkDomain,
}
```

Reply:
trace_polys is an iterator so it only interpolates as .next() is called which is why an iterator is passed to SegmentGenerator rather than a ColMatrix. I'm doing it this way so all the trace polys aren't generated entirely up front but rather only as they're needed.

Reply:
Yep - makes sense!